// pcf-scripts/constants.js
"use strict";
// Copyright (C) Microsoft Corporation. All rights reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SF_SHARED_TEMPLATE = exports.SF_SUPPORT_RESETTABLE_PROPS = exports.SUBSCRIBED_FUNCTIONALITY_SECTION_NODE = exports.WEBPACK_CUSTOMIZATION_FILE_NAME = exports.MAX_BUNDLE_SIZE_IN_MB = exports.INDEX_FILE_NAME = exports.NODE_MODULES = exports.TYPING_FOLDER_NAME = exports.TEMP_NAMESPACE = exports.PROCESSED_MANIFEST_FILE_NAME = exports.PCF_SCRIPTS_PACKAGE_NAME = exports.PACKAGE_JSON_PATH = exports.PACKAGE_LOCK_JSON_PATH = exports.DESIGN_TYPE_FILE_NAME = exports.MANIFEST_TYPE_FILE_NAME = exports.DESIGN_MAPPING_FOLDER_NAME = exports.DESIGN_MAP_INPUT_FILE_NAME = exports.MANIFEST_INPUT_FILE_NAME = exports.PLATFORM_LIBRARY_ELEM_NAME = exports.LIBRARY_ELEM_NAME = exports.DEFAULT_DESIGN_MAP_DIR = exports.DEFAULT_OUT_DIR = exports.DEFAULT_LANGUAGE = exports.DEFAULT_BUILD_MODE = exports.FEATURE_CONFIG_FILE_NAME = exports.CONFIGURATION_FILE_NAME = exports.COMPONENT_FRAMEWORK_INTERNAL_PACKAGE_NAME = exports.COMPONENT_FRAMEWORK_PACKAGE_NAME = exports.COLOR_YELLOW_RESET = exports.COLOR_RED_RESET = exports.CODE_ELEM_NAME = exports.BUNDLE_NAME = exports.BUILD_TOOL_NAME = void 0;
exports.BUILD_TOOL_NAME = 'pac';
exports.BUNDLE_NAME = 'bundle.js';
exports.CODE_ELEM_NAME = 'code';
exports.COLOR_RED_RESET = '\x1b[31m%s\x1b[0m';
exports.COLOR_YELLOW_RESET = '\x1b[33m%s\x1b[0m';
exports.COMPONENT_FRAMEWORK_PACKAGE_NAME = '@types/powerapps-component-framework';
exports.COMPONENT_FRAMEWORK_INTERNAL_PACKAGE_NAME = '@pcf/public-typings';
exports.CONFIGURATION_FILE_NAME = 'pcfconfig.json';
exports.FEATURE_CONFIG_FILE_NAME = 'featureconfig.json';
exports.DEFAULT_BUILD_MODE = 'development';
exports.DEFAULT_LANGUAGE = 'en';
exports.DEFAULT_OUT_DIR = 'out/controls';
exports.DEFAULT_DESIGN_MAP_DIR = 'design-maps/';
exports.LIBRARY_ELEM_NAME = 'library';
exports.PLATFORM_LIBRARY_ELEM_NAME = 'platform-library';
exports.MANIFEST_INPUT_FILE_NAME = 'ControlManifest.Input.xml';
exports.DESIGN_MAP_INPUT_FILE_NAME = 'map.Fluent.xml';
exports.DESIGN_MAPPING_FOLDER_NAME = 'design-maps';
exports.MANIFEST_TYPE_FILE_NAME = 'ManifestTypes.d.ts';
exports.DESIGN_TYPE_FILE_NAME = 'ManifestDesignTypes.d.ts';
exports.PACKAGE_LOCK_JSON_PATH = 'package-lock.json';
exports.PACKAGE_JSON_PATH = 'package.json';
exports.PCF_SCRIPTS_PACKAGE_NAME = 'pcf-scripts';
exports.PROCESSED_MANIFEST_FILE_NAME = 'ControlManifest.xml';
exports.TEMP_NAMESPACE = 'pcf_tools_652ac3f36e1e4bca82eb3c1dc44e6fad';
exports.TYPING_FOLDER_NAME = 'generated';
exports.NODE_MODULES = 'node_modules';
exports.INDEX_FILE_NAME = 'index.ts';
exports.MAX_BUNDLE_SIZE_IN_MB = 5;
exports.WEBPACK_CUSTOMIZATION_FILE_NAME = 'webpack.config.js';
exports.SUBSCRIBED_FUNCTIONALITY_SECTION_NODE = 'subscribed-functionalities';
exports.SF_SUPPORT_RESETTABLE_PROPS = 'supportResettableProperties';
exports.SF_SHARED_TEMPLATE = 'sharedTemplate';
