// pcf-start/generated/telemetry/apptelemetryclient.js
"use strict";
// Copyright (C) Microsoft Corporation. All rights reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuildSource = exports.TelemetryEvent = exports.AppTelemetryClient = exports.setupAppInsightsAndClient = void 0;
var appInsights = require("applicationinsights");
var customtelemetryconfiguration_1 = require("./customtelemetryconfiguration");
var uuid_1 = require("uuid");
var pp_tooling_telemetry_node_1 = require("../pp-tooling-telemetry-node");
function setupAppInsightsAndClient(productName, productVersion, environment, logger) {
    if (!productName)
        throw new Error("productName must be specified.");
    if (!productVersion)
        throw new Error("productVersion must be specified.");
    logger = logger !== null && logger !== void 0 ? logger : console; // Default to global console
    var devLogger = environment.developerMode ? logger : undefined;
    var appInsightsResourceProvider = customtelemetryconfiguration_1.pcfAppInsightsResourceProvider;
    var userSettings = pp_tooling_telemetry_node_1.AppTelemetryConfigUtility.getUserSettingsFromSharedInstall(logger);
    var sessionId = uuid_1.v4();
    devLogger === null || devLogger === void 0 ? void 0 : devLogger.warn("[AppTelemetryClient] SessionId: '" + sessionId + "'. UserId: " + userSettings.uniqueId + "; productVersion: '" + productVersion + "'");
    pp_tooling_telemetry_node_1.AppTelemetryConfigUtility.setupAndStartAppInsights(appInsightsResourceProvider, environment);
    var aiClient = appInsights.defaultClient;
    pp_tooling_telemetry_node_1.AppTelemetryConfigUtility.configureTelemetryClient(aiClient, productName, productVersion, sessionId, environment, userSettings);
    // Placeholder: Add any app-custom telemetry processors
    pp_tooling_telemetry_node_1.AppTelemetryConfigUtility.registerCommonFinalTelemetryProcessors(aiClient, environment, logger);
    return new AppTelemetryClient(aiClient, logger);
}
exports.setupAppInsightsAndClient = setupAppInsightsAndClient;
var AppTelemetryClient = /** @class */ (function () {
    function AppTelemetryClient(aiClient, logger) {
        this._logger = logger;
        this._client = aiClient;
    }
    AppTelemetryClient.prototype.trackEvent = function (telemetry) {
        try {
            this._client.trackEvent(telemetry);
        }
        catch (error) {
            /* don't fail pcf-* due to telemetry errors */
            this.reportSwallowedError("trackEvent", error);
        }
    };
    AppTelemetryClient.prototype.trackException = function (telemetry) {
        try {
            this._client.trackException(telemetry);
        }
        catch (error) {
            /* don't fail pcf-* due to telemetry errors */
            this.reportSwallowedError("trackException", error);
        }
    };
    AppTelemetryClient.prototype.trackRequest = function (telemetry) {
        try {
            this._client.trackRequest(telemetry);
        }
        catch (error) {
            /* don't fail pcf-* due to telemetry errors */
            this.reportSwallowedError("trackRequest", error);
        }
    };
    AppTelemetryClient.prototype.flush = function () {
        try {
            this._client.flush();
        }
        catch (error) {
            /* don't fail pcf-* due to telemetry errors */
            this.reportSwallowedError("flush", error);
        }
    };
    AppTelemetryClient.prototype.reportSwallowedError = function (methodName, error) {
        var _a;
        (_a = this._logger) === null || _a === void 0 ? void 0 : _a.warn("[AppTelemetryClient] unexpected error occured in method '" + methodName + "': " + error);
    };
    return AppTelemetryClient;
}());
exports.AppTelemetryClient = AppTelemetryClient;
var TelemetryEvent;
(function (TelemetryEvent) {
    TelemetryEvent[TelemetryEvent["Start"] = 0] = "Start";
    TelemetryEvent[TelemetryEvent["End"] = 1] = "End";
    TelemetryEvent[TelemetryEvent["StartExecutingVerb"] = 2] = "StartExecutingVerb";
    TelemetryEvent[TelemetryEvent["EndExecutingVerb"] = 3] = "EndExecutingVerb";
})(TelemetryEvent = exports.TelemetryEvent || (exports.TelemetryEvent = {}));
var BuildSource;
(function (BuildSource) {
    BuildSource[BuildSource["VisualStudio"] = 0] = "VisualStudio";
    BuildSource[BuildSource["MSBuild"] = 1] = "MSBuild";
    BuildSource[BuildSource["NPM"] = 2] = "NPM";
})(BuildSource = exports.BuildSource || (exports.BuildSource = {}));
