// fela-bindings/lib/FelaComponentFactory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = FelaComponentFactory;

var _fela = require('fela');

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function FelaComponentFactory(createElement, RendererContext, FelaTheme) {
  function FelaComponent(_ref) {
    var children = _ref.children,
        _ref$as = _ref.as,
        as = _ref$as === undefined ? 'div' : _ref$as,
        style = _ref.style,
        otherProps = _objectWithoutProperties(_ref, ['children', 'as', 'style']);

    var renderFn = function renderFn(renderer) {
      if (renderer.devMode && style == null) {
        // eslint-disable-next-line no-console
        console.warn('"FelaComponent" is being rendered without a style prop\nIf all you need is access to theme, try using "FelaTheme" or the "useFela" hook instead');
      }

      return createElement(FelaTheme, undefined, function (theme) {
        // TODO: could optimize perf by not calling combineRules if not necessary
        var renderedRule = renderer.renderRule((0, _fela.combineRules)(style), _extends({}, otherProps, {
          theme: theme
        }));

        if (children instanceof Function) {
          return children({
            className: !renderer.isNativeRenderer && renderedRule,
            style: renderer.isNativeRenderer && renderedRule,
            theme: theme,
            as: as
          });
        }

        return createElement(as, {
          className: !renderer.isNativeRenderer && renderedRule,
          style: renderer.isNativeRenderer && renderedRule
        }, children);
      });
    };

    return createElement(RendererContext.Consumer, undefined, renderFn);
  }

  return FelaComponent;
}
