// @microsoft/eslint-plugin-power-apps/lib/rules/avoid-ui-refreshribbon.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const ruleUtils = require("../utils/rule-utils");

//------------------------------------------------------------------------------
// Constants
//------------------------------------------------------------------------------

const MESSAGE_ID = "refreshRibbonFound";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "Do not use ui.refreshRibbon during evaluation of ribbon rules. This triggers duplicated network calls from ribbon and degrades performance. Use promises and asynchronous patterns.",
            category: "Performance",
            recommended: true,
            url: "https://docs.microsoft.com/power-apps/maker/data-platform/powerapps-checker/rules/web/avoid-ui-refreshribbon"
        },

        messages: {
            [MESSAGE_ID]: " {{arg0}} calls ui.refreshRibbon. This will result in performance degradations when used inside Ribbon rules. Please avoid this anti-pattern to hide/show commands."
        }
    },

    create(context) {

        const sourceCode = context.getSourceCode();
        
        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------

        return {
            CallExpression: function(node){
                const snippet = sourceCode.getText(node);
                const callee = node.callee;
                if(callee.object 
                    && callee.object.property 
                    && callee.object.property.name == 'ui'
                    && callee.property
                    && callee.property.name == 'refreshRibbon'
                ){
                    ruleUtils.reportAccordingly(
                            context,
                            node,
                            MESSAGE_ID,
                            [snippet],
                            snippet
                    );
                }

            }
            
        }
    }
};
