// @microsoft/eslint-plugin-power-apps/lib/rules/use-getsecurityroleprivilegesinfo.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const ruleUtils = require("../utils/rule-utils");

//------------------------------------------------------------------------------
// Constants
//------------------------------------------------------------------------------

const MESSAGE_ID = "use_getSecurityRolePrivilegesInfo";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        type: "suggestion",

        docs: {
            description: "Consider using the Xrm.Utility.getGlobalContext().userSettings.getSecurityRolePrivilegesInfo() API when working with security role privileges in JavaScript.",
            category: "Performance",
            recommended: true,
            url: "https://docs.microsoft.com/power-apps/maker/data-platform/powerapps-checker/rules/web/use-getsecurityroleprivilegesinfo"
        },

        messages: {
            [MESSAGE_ID]: " {{arg0}} calls userSettings.securityRolePrivileges. This returns only privilege GUIDs which are difficult to work with. You might be making additional network calls to get further details about the privilege GUID and that degrades performance. Use userSettings.getSecurityRolePrivilegesInfo() instead which gives you more details about each security role privilege."
        }
    },

    create(context) {

        const sourceCode = context.getSourceCode();
        
        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------

        return {
            MemberExpression: function(node){
                const snippet = sourceCode.getText(node);
                if(node.object 
                    && node.object.property 
                    && node.object.property.name == 'userSettings'
                    && node.property
                    && node.property.name == 'securityRolePrivileges'
                ){
                    ruleUtils.reportAccordingly(
                            context,
                            node,
                            MESSAGE_ID,
                            [snippet],
                            snippet
                    );
                }

            }
            
        }
    }
};
