// pcf-scripts/bin/pcf-scripts.js
#!/usr/bin/env node
"use strict";
// Copyright (C) Microsoft Corporation. All rights reserved.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const yargs_1 = require("yargs");
const fs = require("fs");
const path = require("path");
const constants = require("../constants");
const taskRunner_1 = require("../taskRunner");
const taskGroup_1 = require("../taskGroup");
const PackageVersionUtils_1 = require("../generated/PackageVersionUtils");
const apptelemetryclient_1 = require("../generated/telemetry/apptelemetryclient");
const pp_tooling_telemetry_node_1 = require("../generated/pp-tooling-telemetry-node");
RunAsync();
function RunAsync() {
    return __awaiter(this, void 0, void 0, function* () {
        const telemetryEnv = pp_tooling_telemetry_node_1.AppTelemetryConfigUtility.createGlobalTelemetryEnvironment();
        const version = (0, PackageVersionUtils_1.getPackageVersion)(constants.PCF_SCRIPTS_PACKAGE_NAME, process.cwd()) || '0.0.1';
        const telemetryClient = (0, apptelemetryclient_1.setupAppInsightsAndClient)('pcf-scripts', version, telemetryEnv);
        let result = {};
        let resultCode = 0;
        const startTime = Date.now();
        let buildSource;
        if (yargs_1.argv['buildSource'] === apptelemetryclient_1.BuildSource[apptelemetryclient_1.BuildSource.VisualStudio]) {
            buildSource = apptelemetryclient_1.BuildSource.VisualStudio;
        }
        else if (yargs_1.argv['buildSource'] === apptelemetryclient_1.BuildSource[apptelemetryclient_1.BuildSource.MSBuild]) {
            buildSource = apptelemetryclient_1.BuildSource.MSBuild;
        }
        else {
            buildSource = apptelemetryclient_1.BuildSource.NPM;
        }
        let sessionStartProperties = Object.assign({ 'taskGroup': yargs_1.argv._[0], 'buildSource': apptelemetryclient_1.BuildSource[buildSource] }, pp_tooling_telemetry_node_1.AppTelemetryUtility.createCommonAppStartProperties(telemetryEnv));
        try {
            telemetryClient.trackEvent({
                name: apptelemetryclient_1.TelemetryEvent[apptelemetryclient_1.TelemetryEvent.Start],
                properties: Object.assign({}, sessionStartProperties)
            });
            if (!yargs_1.argv._ || yargs_1.argv._.length === 0 || !Object.keys(taskGroup_1.standardTaskGroups).includes(yargs_1.argv._[0].toString())) {
                printUsage();
                resultCode = 400;
                return;
            }
            const fileConfig = fs.readFileSync(path.resolve('.', constants.CONFIGURATION_FILE_NAME));
            const config = JSON.parse(fileConfig.toString());
            const taskRunner = new taskRunner_1.TaskRunner(yargs_1.argv, config);
            result = yield taskRunner.run(taskGroup_1.standardTaskGroups);
            resultCode = 200;
        }
        catch (e) {
            resultCode = 500;
            console.error(`[pcf-scripts] [Error] encountered unexpected error:\n${e}`);
            telemetryClient.trackException({ exception: e });
        }
        finally {
            const isSuccessful = resultCode === 200;
            process.exitCode = isSuccessful ? 0 : 1;
            let customMeasurement = {};
            if (isSuccessful && !!(result === null || result === void 0 ? void 0 : result.compileTime)) {
                customMeasurement['compileTimeMs'] = result.compileTime;
            }
            telemetryClient.trackEvent({ name: apptelemetryclient_1.TelemetryEvent[apptelemetryclient_1.TelemetryEvent.End], measurements: customMeasurement });
            telemetryClient.trackRequest({
                name: 'Session',
                url: '',
                duration: Date.now() - startTime,
                resultCode: resultCode,
                success: isSuccessful,
                properties: Object.assign({}, sessionStartProperties)
            });
            telemetryClient.flush();
        }
    });
}
function printUsage() {
    console.log('Unsupported command');
    console.log(`Usage: pcf-scripts ${Object.keys(taskGroup_1.standardTaskGroups).map(verb => '[' + verb + ']').join(' ')}`);
}
